src/python/pants/backend/jvm/tasks/jvm_compile/jvm_classpath_publisher.py

```diff
@@ -35,7 +37,9 @@ def execute(self):
       # Safely create e.g. dist/export-classpath/manifest.jar
       safe_classpath(classpath, basedir, "manifest.jar")
```

Comment:
Should `manifest-jar-only` need to respect the option as well? so filtering `classpath` maybe be required.

Reply:
Looking at the definition of `ClasspathUtil.classpath()`, it doesn't appear to be possible to restrict to internal only? It seems to only be a concept when creating a "canonical" classpath: https://github.com/pantsbuild/pants/blob/cd92c65aeb3dfdcee3e0946f2b68a301ef2f4541/src/python/pants/backend/jvm/tasks/classpath_util.py#L64

Reply:
It could be implemented if this was a necessary use case. I'm not fully clear on how the manifest jar option is used.

Reply:
Interesting, I didn't know about `--manifest-jar-only`. An example of a use case for that flag _today_ would be using this in order to access binary jar dependencies, since we can just open the manifest in the jar, get the Class-Path attribute, split on spaces, and use the resultant relative paths to access jars in the `.pants.d` directory. In fact, I think I'll switch to using `--manifest-jar-only` for now, at least until we deploy a build of Pants with this fix.

On the one hand, it's nice for consistency to have `--internal-classpath-only` work analogously whether you use `--manifest-jar-only` or not. On the other hand, is there a valid use case for `--internal-classpath-only` in the long term, or is it only useful for transitioning to the new behavior? And on the third hand, is the old behavior something that anyone even relies on today?

One thing I can say is if you do make `--internal-classpath-only` affect `--manifest-jar-only`, you'd have to make it default to `false` if  `--manifest-jar-only` is specified (so as not to break manifest jar files, which currently do include the full classpath) but default to `true` otherwise. That sounds complicated, and not being familiar with Pants internals, I don't know if there's a good way to make it work like that and be correctly documented in `--help`.

Reply:
There are a lo of ways to go with this, including the use of the fantastic deprecation utilities in https://github.com/pantsbuild/pants/blob/master/src/python/pants/base/deprecated.py, but I'm actually leaning towards changing the name of the `--manifest-jar-only` option, since I feel like it has a distinctly different meaning that the other two new `--*-only` flags in this PR and in #8605.

I'm thinking we can introduce a new flag `--manifest-jar` which has the behavior we'd like when composed with `--internal-classpath-only`, and just deprecate `--manifest-jar-only`. @wisechengyi thoughts on that approach?

Reply:
Implemented the above, in the process adding testing for the `--manifest-jar` option for the first time (which is why I was reluctant to take the time to implement it earlier and would have preferred to avoid introducing it, but it's always good to add testing).